### PR TITLE
DBZ-1336 Support Postgres LTREE columns

### PR DESCRIPTION
**Pull request by mincong-h**
https://issues.jboss.org/browse/DBZ-1336

This PR aims to support PostgreSQL [LTREE](https://www.postgresql.org/docs/current/ltree.html) column and LTREE array column. This is still a draft because I need your guidance to move forward 🙏

### CODE REVIEW

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresValueConverter.java [hunk 1]
```diff
@@ -254,6 +255,9 @@ else if (oidValue == typeRegistry.geometryArrayOid()) {
                 else if (oidValue == typeRegistry.hstoreOid()) {
                     return hstoreSchema();
                 }
+                else if (oidValue == typeRegistry.ltreeOid()) {
+                    return SchemaBuilder.string().optional();
```
Comment (mincong-h) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresValueConverter.java]: Should I use a dedicated schema, such as one with logical name called "io.debezium.data.Ltree"?

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresValueConverter.java [hunk 2]
```diff
@@ -486,6 +499,29 @@ protected Object convertHStore(Column column, Field fieldDefn, Object data, HSto
             return convertHstoreToMap(column, fieldDefn, data);
         }
 
+    private Object convertLtree(Column column, Field fieldDefn, Object data) {
+        return convertValue(column, fieldDefn, data, "", r -> {
+            if (data instanceof String) {
+                r.deliver(data);
```
Comment (mincong-h) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresValueConverter.java]: For now, the data is delivered as `java.lang.String`. Should we consider using a real Java representation, such as `Ltree`? This class can contain functions shown in PostgreSQL, see [PostgreSQL Documentation - F.14. ltree Functions](https://www.postgresql.org/docs/current/ltree.html#LTREE-FUNC-TABLE). What I imagine is:

```java
r.deliver(Ltree.parse(data));
```

or

```
r.deliver(new Ltree(data));
```

where "data" is the label of the ltree.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresValueConverter.java [hunk 3]
```diff
@@ -486,6 +499,29 @@ protected Object convertHStore(Column column, Field fieldDefn, Object data, HSto
...
+            }
+            else if (data instanceof PGobject) {
+                r.deliver(data.toString());
+            }
+        });
+    }
+
+    private Object convertLtreeArray(Column column, Field fieldDefn, Object data) {
```
Comment (mincong-h) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresValueConverter.java]: From the failing test results, I can see that I didn't handle the array correctly. But I don't know what should be the right way...

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresValueConverter.java (hunk 1 above)
Comment (gunnarmorling) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresValueConverter.java]: Yes, please. Similar to hstore where I think we do the same (or JSON). These semantical schemas allow consumers to specifically react to such values, e.g. to derive a corresponding schema in a downstream database.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresValueConverter.java (hunk 2 above)
Comment (gunnarmorling) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresValueConverter.java]: I don't think that's needed. In the end, the value would be propagated as a string anyways.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresValueConverter.java (hunk 3 above)
Comment (gunnarmorling) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresValueConverter.java]: What test failure do you get exactly?

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresValueConverter.java (hunk 1 above)
Comment (mincong-h) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresValueConverter.java]: The problem I encountered last weekend was that some schema don't have the semantical schema while the others do. I believe they are stored here: https://github.com/debezium/debezium/tree/master/debezium-core/src/main/java/io/debezium/data

The ones having semantical schemas (excerpt):

- Json
- Uuid
- Xml

The ones without semantical schemas (excerpt):

- Hstore ([declared as map](https://github.com/debezium/debezium/blob/master/debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresValueConverter.java#L283-L294))
- Citext ([declared as string](https://github.com/debezium/debezium/blob/master/debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresValueConverter.java#L248-L250))

What confuse me most is that those without semantical schemas are the ones declared as extensions in [init_database.ddl](https://github.com/debezium/debezium/blob/2a065686bd8df50ae5c427d2847e0008a6b5f256/debezium-connector-postgres/src/test/resources/init_database.ddl). So if I need to align with them, I should not declare a semantical schema.

**Comment by mincong-h** [comment]
Hmm, still one failure left... https://travis-ci.org/debezium/debezium/jobs/596977007#L19651

```
RecordsStreamProducerIT.testEmptyChangesProducesHeartbeat:1346->assertHeartBeatRecordInserted:1521 expected:<[__debezium-heartbeat.test_server]> but was:<[test_server.public.test_table]>
```

I can reproduce locally, but not sure how my changes impacted it...

**Comment by gunnarmorling** [comment]
Hey @mincong-h, sorry for the delay, I was travelling during the last week. Is it good from your side for another review round?

**Comment by mincong-h** [comment]
Hi, Gunnar. Thanks for your feedback. Yes, I'm ready for another review round. Btw, all tests passed now.

**Comment by gunnarmorling** [comment]
Excellent, thanks! I'll check it out asap.

>

**Pull request closed without merging**